ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt

```diff
@@ -514,3 +514,84 @@ suspend fun NumberNDArrayTFJS.batchNorm(
         NumberNDArrayTFJS(batchNorm(tfjsArray, bcastMean.tfjsArray, bcastVar.tfjsArray, bcastBias.tfjsArray, bcastScale.tfjsArray, epsilon))
     }
 }
+
+suspend fun NumberNDArrayTFJS.conv(
+    w: NumberNDArrayTFJS,
+    b: NumberNDArrayTFJS?,
+    groups: Int,
+    autoPad: String,
+    pads: IntArray?,
+    strides: IntArray?,
+    dilations: IntArray?
+) : NumberNDArrayTFJS {
+    val effectiveStrides: Array<Number> = (strides ?: IntArray(rank - 2) { 1 }).map { it }.toTypedArray()
+    val effectiveDilations: Array<Number> = (dilations ?: IntArray(rank - 2) { 1 }).map { it }.toTypedArray()
+    val padding: Any = convertOnnxPaddingToTfjs(autoPad, pads, rank - 2)
+    val convOutputs = mutableListOf<ArrayTFJS>()
+    val inputChannels = shape[1]
+    val channelsPerGroup = inputChannels / groups
+
+    val (transposeArr, convFunction) = when (rank) {
+        3 -> intArrayOf(2, 1, 0) to { i: ArrayTFJS, f: ArrayTFJS, s: Array<Number>, p: Any, d: Array<Number> ->
+            conv1d(i, f, s.first(), p, "NCW", d.first())
+        }
+        4 -> intArrayOf(2, 3, 1, 0) to { i: ArrayTFJS, f: ArrayTFJS, s: Array<Number>, p: Any, d: Array<Number> ->
+            conv2d(i, f, s, p, "NCHW", d)
+        }
+        5 -> intArrayOf(2, 3, 4, 1, 0) to { i: ArrayTFJS, f: ArrayTFJS, s: Array<Number>, p: Any, d: Array<Number> ->
+            conv3d(i, f, s, p, "NCDHW", d)
+        }
+        else -> error("Unsupported tensor rank for convolution: $rank")
```

Comment:
I think we need to specify which ranks are supported in this message

Reply:
Fixed in [2cd4c8e](https://github.com/JetBrains-Research/kinference/pull/160/commits/2cd4c8e2b833a65e64cece8aaf487e8636f0bdeb)